Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m

```diff
@@ -0,0 +1,356 @@
+        PWI4D = ControlIm - LabelIm;
+        
+        % Skip every other value in the vectors as they were stored for both control and label images 
+        x.Q.EchoTime_PWI4D = x.Q.EchoTime(1:2:end);
+        x.Q.InitialPLD_PWI4D = x.Q.Initial_PLD(1:2:end);
+        x.Q.LabelingDuration_PWI4D = x.Q.LabelingDuration(1:2:end);
+    else % the same but then without subtraction
+        PWI4D = ASL4D;
+    end
+
+end
+   
+if bCreatePWI3D
+    %% 5. Create PWI3D
+    
+    % After averaging across PLDs, we'll obtain these unique PLDs+LD combinations
+    % indexAverage_PLD_LabDur lists for each original position to where it should be averaged
+    [~, ~, iUnique_TE_PLD_LabDur] = unique([x.Q.EchoTime_PWI4D(:), x.Q.InitialPLD_PWI4D(:), x.Q.LabelingDuration_PWI4D(:)], 'stable', 'rows');
+    
+    % MultiPLD-multiLabDur PWI3D after averaging
+    for iTE_PLD_LabDur = 1:max(iUnique_TE_PLD_LabDur)
+        indicesAre = iUnique_TE_PLD_LabDur == iTE_PLD_LabDur;
+        PWI3D(:, :, :, iTE_PLD_LabDur) = xASL_stat_MeanNan(PWI4D(:, :, :, indicesAre), 4); % Averaged PWI4D
+        x.Q.EchoTime_PWI3D(iTE_PLD_LabDur) = xASL_stat_MeanNan(x.Q.EchoTime_PWI4D(indicesAre));
+        x.Q.InitialPLD_PWI3D(iTE_PLD_LabDur) = xASL_stat_MeanNan(x.Q.InitialPLD_PWI4D(indicesAre));
+        x.Q.LabelingDuration_PWI3D(iTE_PLD_LabDur) = xASL_stat_MeanNan(x.Q.LabelingDuration_PWI4D(indicesAre));
+    end
+end
+
+%% 6. Create PWI
+% We create a dummy CBF image for registration purposes
+% The earliest echo, the latest PLD and the longest labeling duration
+% are the best for this, having most SNR, CBF-weighting, and SNR,
+% respectively. 
+% But from a single volume, we have little signal.
+% So we take a weighted average accordingly
+
+%% This would be taking a single volume only
+% % The earliest echo has the most SNR for perfusion-weighting
+% iMinTE = x.Q.EchoTime_PWI3D == min(x.Q.EchoTime_PWI3D(:));
+% % The latest PLD has the most perfusion-weighting
+% iMaxPLD = x.Q.InitialPLD_PWI3D == max(x.Q.InitialPLD_PWI3D(:));
+% % The longest labeling duration has the most SNR and most perfusion-weighting
+% iMaxLabelingDuration = x.Q.LabelingDuration_PWI3D == max(x.Q.LabelingDuration_PWI3D(:));
+% % We take the index that has all these        
+% i3D = iMinTE & iMaxPLD & iMaxLabelingDuration;
+% if isempty(i3D) || sum(i3D)~=1
+%     error('Illegal index for PWI image');
+% end
+% PWI = PWI3D(:, :, :, i3D);
+% x.Q.EchoTime_PWI = min(x.Q.EchoTime_PWI3D(:));
+% x.Q.initialPLD_PWI = max(x.Q.InitialPLD_PWI3D(:));
+% x.Q.LabelingDuration_PWI = max(x.Q.LabelingDuration_PWI3D(:));
+
+%% PM: do a weighted average here?
+
+%% Current quick&dirty fix was the average of all volumes with the lowest echo time
+iMinTE = x.Q.EchoTime_PWI3D == min(x.Q.EchoTime_PWI3D(:));
+
+PWI = xASL_stat_MeanNan(PWI3D(:, :, :, iMinTE), 4);
+x.Q.EchoTime_PWI = min(x.Q.EchoTime_PWI3D(:));
+x.Q.initialPLD_PWI = mean(x.Q.InitialPLD_PWI3D(:)); %% Jan is this correct?
```

Comment:
You will anyway change this to take the best PLD, so pointless to comment.

Reply:
True. See above. Unless we decide to do this in another issue.

Reply:
This still needs to be decided.

Reply:
I wanted to ask if this is correct for the time being. We had this average for the time being.

I would vote for a weighted average in our next issue. You might argue that this needs testing, but the same can be said for only taking the last PLD or the average of all PLDs, that also needs to be tested.
The weighted average is essentially an in-between option between averaging all and only taking the correct/latest PLD.

Reply:
OK. For now, for QC and registration, an average is fine. 
I'm fine to do this without more testing for now.

Weighted average is the best idea. Because taking the last might not work if the last has high SNR - cause then we could take several weighted lasts.

So yes, we should move this to #1543 and remove that comment from the code.

Reply:
Done in 93abfd7f00898dd99847d5da61daeca7d94aa165